src/backend/optimizer/path/allpaths.c

```diff
@@ -2953,6 +2953,16 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 
 	}
 
+	/*
+	 * Inlining the non-plain SELECT CTEs (like INSERT/UPDATE/DELETE) leads to
+	 * redundant or duplicated operations whenever the CTE is referenced
+	 * multiple times within the query. Therefore, the best option is to
+	 * materialize such CTEs by utilizing the existing SharedInputScan node.
+	 * This idea is similar to vanilla PostgreSQL approach.
+	 */
+	if (subquery->commandType != CMD_SELECT)
+		is_shared = true;
```

Comment:
Is it possible that the code a few lines below will cancel sharing again?

Reply:
No, contain_outer_selfref() checks whether cte contains **outer** self-refernce to recursive cte. 
```
explain (verbose, costs off)
with recursive x(a) as
  ((values ('a'), ('b'))
   union all
   (with z as  materialized (select * from x)
    select z.a || z1.a as a from z cross join z as z1
    where length(z.a || z1.a) < 5))
select * from x;
                    QUERY PLAN                     
---------------------------------------------------
 Recursive Union
   ->  Values Scan on "*VALUES*"
         Output: "*VALUES*".column1
   ->  Nested Loop
         Output: (x.a || x_1.a)
         Join Filter: (length((x.a || x_1.a)) < 5)
         ->  WorkTable Scan on x
               Output: x.a
         ->  WorkTable Scan on x x_1
               Output: x_1.a
 Optimizer: Postgres-based planner
 Settings: optimizer = 'off'
(12 rows)
```
```
The condition below prevents `z` from being shared because it contains self-refernce rte `x`.

In case of modifying DML we cannot achive the same structure:
explain (costs off)
with recursive cte as (
    select 1 as i from t1 where t1.j = 2
    union all
    select foo.j + 1 from (with cte2 as (update t1 set i = cte.i from cte where cte.i=t1.j) select * from cte2) foo where foo.i=1)
 select * from cte;
 ERROR:  WITH clause containing a data-modifying statement must be at the top level
LINE 5:     select foo.j + 1 from (with cte2 as (update t1 set i = c...
``` 

This example is also valid (cte3 does not contain outer self refences):
```
explain (costs off)
with recursive cte as (
    select 1 as i from t1 where t1.j = 2
    union all
    select foo.j + 1 from (with cte2 as (select * from cte3 join cte using (i) ) select * from cte2) foo where foo.i=1),
cte3 as (insert into t1 select i, i * 100 from generate_series(1,5) i returning *)
 select * from cte;

                                                 QUERY PLAN                                                 
------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Recursive Union
         ->  Seq Scan on t1
               Filter: (j = 2)
         ->  Subquery Scan on foo
               ->  Subquery Scan on cte2
                     ->  Nested Loop
                           ->  WorkTable Scan on cte
                                 Filter: (i = 1)
                           ->  Materialize
                                 ->  Broadcast Motion 3:3  (slice2; segments: 3)
                                       ->  Subquery Scan on cte3
                                             Filter: (cte3.i = 1)
                                             ->  Shared Scan (share slice:id 2:0)
                                                   ->  Insert on t1 t1_1
                                                         ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                               Hash Key: i.i
                                                               ->  Function Scan on generate_series i
 Optimizer: Postgres-based planner
(19 rows)

```

Reply:
> No

Then why do we need to call the contain_outer_selfref function again if we know for sure that it will return false?

Reply:
what do you suggest, move the code after this condition?

Reply:
may be add else? like
```suggestion
	if (subquery->commandType != CMD_SELECT)
		is_shared = true;
	else if (is_shared && contain_outer_selfref(cte->ctequery))
		is_shared = false;
```